src/detect-transform-base64.c

```diff
@@ -0,0 +1,333 @@
+            DetectTransformFromBase64DecodeRegisterTests;
+#endif
+    sigmatch_table[DETECT_TRANSFORM_FROM_BASE64].flags |= SIGMATCH_OPTIONAL_OPT;
+}
+
+static void DetectTransformFromBase64DecodeFree(DetectEngineCtx *de_ctx, void *ptr)
+{
+    ScTransformBase64Free(ptr);
+}
+
+static DetectTransformFromBase64Data *DetectTransformFromBase64DecodeParse(const char *str)
+{
+    DetectTransformFromBase64Data *tbd = ScTransformBase64Parse(str);
+    if (tbd == NULL) {
+        SCLogError("invalid transform_base64 values");
+    }
+    return tbd;
+}
+
+/**
+ *  \internal
+ *  \brief Base64 decode the input buffer
+ *  \param det_ctx detection engine ctx
+ *  \param s signature
+ *  \param opts_str transform options, if any
+ *  \retval 0 No decode
+ *  \retval >0 Decoded byte count
+ */
+static int DetectTransformFromBase64DecodeSetup(
+        DetectEngineCtx *de_ctx, Signature *s, const char *opts_str)
+{
+    int r = -1;
+
+    SCEnter();
+
+    DetectTransformFromBase64Data *b64d = DetectTransformFromBase64DecodeParse(opts_str);
+    if (b64d == NULL)
+        SCReturnInt(r);
+
+    if (b64d->flags & DETECT_TRANSFORM_BASE64_FLAG_OFFSET_VAR) {
+        SCLogError("offset value must be a value, not a variable name");
+        goto exit_path;
+    }
+
+    if (b64d->flags & DETECT_TRANSFORM_BASE64_FLAG_NBYTES_VAR) {
+        SCLogError("byte value must be a value, not a variable name");
+        goto exit_path;
+    }
+
+    r = DetectSignatureAddTransform(s, DETECT_TRANSFORM_FROM_BASE64, b64d);
+
+exit_path:
+    if (r != 0)
+        DetectTransformFromBase64DecodeFree(de_ctx, b64d);
+    SCReturnInt(r);
+}
+
+static void TransformFromBase64Decode(InspectionBuffer *buffer, void *options)
+{
+    DetectTransformFromBase64Data *b64d = options;
+    const uint8_t *input = buffer->inspect;
+    const uint32_t input_len = buffer->inspect_len;
+    /*
+     * The output buffer is larger than needed. On average,
+     * a base64 encoded string is 75% of the unencoded
+     * string.
+     */
+    uint8_t output[input_len];
+    uint32_t decode_length = input_len;
+
+    DetectBase64Mode mode = b64d->mode;
+    uint32_t offset = b64d->offset;
+    uint32_t nbytes = b64d->nbytes;
+
+    if (offset) {
+        if (offset > input_len) {
+            SCLogDebug("offset %d exceeds length %d; returning", offset, input_len);
+            return;
+        }
+        input += offset;
+        decode_length -= offset;
+    }
+
+    if (nbytes) {
+        if (nbytes > decode_length) {
+            SCLogDebug("byte count %d plus offset %d exceeds length %d; returning", nbytes, offset,
+                    input_len);
+            return;
+        }
+        decode_length = nbytes;
+    }
+
+    // PrintRawDataFp(stdout, input, input_len);
+    uint32_t decoded_length;
+    uint32_t consumed;
+    Base64Ecode code =
+            DecodeBase64(output, input_len, input, decode_length, &consumed, &decoded_length, mode);
```

Comment:
`mode` should be converted to `Base64Mode` like 0 is relax instead of RFC4648 !!! Is it not ?